File: Pastura/PasturaTests/Engine/Phases/VoteHandlerTests.swift
import Testing

@testable import Pastura

@Suite(.timeLimit(.minutes(1)))
struct VoteHandlerTests {
  let handler = VoteHandler()

  @Test func collectsVotesFromAllAgents() async throws {
    let mock = MockLLMService(responses: [
      #"{"vote": "Bob", "reason": "suspicious"}"#,
      #"{"vote": "Alice", "reason": "quiet"}"#,
      #"{"vote": "Alice", "reason": "weird"}"#
    ])
    try await mock.loadModel()

    let scenario = makeTestScenario(
      phases: [
        Phase(
          type: .vote, prompt: "Vote!", outputSchema: ["vote": "string", "reason": "string"],
          excludeSelf: true)
      ]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    #expect(state.voteResults["Alice"] == 2)
    #expect(state.voteResults["Bob"] == 1)
    #expect(mock.generateCallCount == 3)
  }

  @Test func emitsVoteResultsEvent() async throws {
    let mock = MockLLMService(responses: [
      #"{"vote": "Bob"}"#,
      #"{"vote": "Alice"}"#
    ])
    try await mock.loadModel()

    let scenario = makeTestScenario(
      agentNames: ["Alice", "Bob"],
      phases: [Phase(type: .vote, prompt: "Vote!", outputSchema: ["vote": "string"])]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    let voteEvents = collector.events.compactMap { event -> ([String: String], [String: Int])? in
      if case .voteResults(let votes, let tallies) = event { return (votes, tallies) }
      return nil
    }
    #expect(voteEvents.count == 1)
    #expect(voteEvents[0].0["Alice"] == "Bob")
    #expect(voteEvents[0].0["Bob"] == "Alice")
  }

  @Test func skipsEliminatedAgents() async throws {
    let mock = MockLLMService(responses: [
      #"{"vote": "Charlie"}"#,
      #"{"vote": "Alice"}"#
    ])
    try await mock.loadModel()

    let scenario = makeTestScenario(
      phases: [Phase(type: .vote, prompt: "Vote!", outputSchema: ["vote": "string"])]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    state.eliminated["Bob"] = true
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    #expect(mock.generateCallCount == 2)
  }

  @Test func populatesVoteResultsStateVariable() async throws {
    // Key must be "vote_results" (plural) to match the {vote_results} placeholder
    // documented in PhaseEditorSheet and used by the word_wolf preset.
    let mock = MockLLMService(responses: [
      #"{"vote": "Bob"}"#,
      #"{"vote": "Alice"}"#,
      #"{"vote": "Alice"}"#
    ])
    try await mock.loadModel()

    let scenario = makeTestScenario(
      phases: [Phase(type: .vote, prompt: "Vote!", outputSchema: ["vote": "string"])]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    #expect(state.variables["vote_results"] == #"{"Alice": 2, "Bob": 1}"#)
    #expect(state.variables["vote_result"] == nil)
  }

  @Test func dropsInvalidVotesFromTally() async throws {
    // Regression for #524. Three voters (exclude_self default true):
    //   Alice → "Alice"   self-vote, invalid (Alice ∉ her candidates)
    //   Bob   → "Ghost"   hallucinated name, invalid
    //   Charlie → "Bob"   valid
    // Reverting the drop would put Alice (self) and Ghost back into the
    // tally — this test fails in that case.
    let mock = MockLLMService(responses: [
      #"{"vote": "Alice"}"#,
      #"{"vote": "Ghost"}"#,
      #"{"vote": "Bob"}"#
    ])
    try await mock.loadModel()

    let scenario = makeTestScenario(
      agentNames: ["Alice", "Bob", "Charlie"],
      phases: [Phase(type: .vote, prompt: "Vote!", outputSchema: ["vote": "string"])]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    // Only Charlie's valid vote is tallied.
    #expect(state.voteResults == ["Bob": 1])
    #expect(state.voteResults["Alice"] == nil)  // self-vote dropped
    #expect(state.voteResults["Ghost"] == nil)  // hallucinated dropped

    // Divergence is intentional: the raw votes stay visible in the
    // voteResults event even though they are absent from the tally.
    let voteEvent = collector.events.compactMap { event -> [String: String]? in
      if case .voteResults(let votes, _) = event { return votes }
      return nil
    }.first
    let votes = try #require(voteEvent)
    #expect(votes["Alice"] == "Alice")  // self-vote preserved in votes map
    #expect(votes["Bob"] == "Ghost")  // hallucinated preserved in votes map
    #expect(votes["Charlie"] == "Bob")
  }

  // MARK: - Per-voter enumeration schema (#524)

  /// Extract the `vote` field's kind from the schema captured on the Nth
  /// `generate` call (one per voter, in persona order).
  private func voteKind(
    _ mock: MockLLMService, call index: Int
  ) throws -> OutputSchema.Kind {
    let schema = try #require(mock.capturedSchemas[index])
    let voteField = try #require(schema.fields.first { $0.name == "vote" })
    return voteField.kind
  }

  @Test func voteSchemaConstrainsVoteToCandidates() async throws {
    // exclude_self (default true) → each voter's candidate list omits self.
    // The schema passed for voter Alice must enumerate [Bob, Charlie].
    let mock = MockLLMService(responses: [
      #"{"vote": "Bob"}"#,
      #"{"vote": "Charlie"}"#,
      #"{"vote": "Bob"}"#
    ])
    try await mock.loadModel()

    let scenario = makeTestScenario(
      agentNames: ["Alice", "Bob", "Charlie"],
      phases: [Phase(type: .vote, prompt: "Vote!", outputSchema: ["vote": "string"])]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    // Alice (call 0) cannot vote herself; Charlie (call 2) cannot vote himself.
    #expect(try voteKind(mock, call: 0) == .enumeration(["Bob", "Charlie"]))
    #expect(try voteKind(mock, call: 2) == .enumeration(["Alice", "Bob"]))
  }

  @Test func voteSchemaConstrainsCJKCandidates() async throws {
    // The grammar enumeration must carry CJK persona names verbatim — this
    // is the scenario-factory's primary workload (#524 was observed on
    // Japanese-named personas).
    let mock = MockLLMService(responses: [
      #"{"vote": "博士マッド野"}"#,
      #"{"vote": "体験者ミラクル子"}"#,
      #"{"vote": "開き直りマコ"}"#
    ])
    try await mock.loadModel()

    let scenario = makeTestScenario(
      agentNames: ["開き直りマコ", "博士マッド野", "体験者ミラクル子"],
      phases: [Phase(type: .vote, prompt: "Vote!", outputSchema: ["vote": "string"])]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    #expect(
      try voteKind(mock, call: 0) == .enumeration(["博士マッド野", "体験者ミラクル子"]))
    // Assert a second voter too, so an ordering / filter regression on
    // multi-byte names can't slip through on a single-call lock.
    #expect(
      try voteKind(mock, call: 2) == .enumeration(["開き直りマコ", "博士マッド野"]))
  }

  @Test func voteSchemaFallsBackToStringForUnsafeCandidate() async throws {
    // A persona name containing a GBNF-hostile char (`"`) would make the
    // enumeration grammar throw and abort the run. The voter whose
    // candidate list includes that name must degrade to `.string`.
    let unsafeName = #"Ali"ce"#
    let mock = MockLLMService(responses: [
      #"{"vote": "Bob"}"#,
      #"{"vote": "Charlie"}"#,
      #"{"vote": "Bob"}"#
    ])
    try await mock.loadModel()

    let scenario = makeTestScenario(
      agentNames: [unsafeName, "Bob", "Charlie"],
      phases: [Phase(type: .vote, prompt: "Vote!", outputSchema: ["vote": "string"])]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    // Voter 0 (unsafe name) excludes self → candidates [Bob, Charlie], all
    // safe → still enumerated.
    #expect(try voteKind(mock, call: 0) == .enumeration(["Bob", "Charlie"]))
    // Voter 1 (Bob) → candidates [Ali"ce, Charlie] include the unsafe name
    // → fall back to `.string` rather than abort.
    #expect(try voteKind(mock, call: 1) == .string)
  }

  // MARK: - simulationLanguage override (ADR-010 Step E)

  @Test func voteHonorsSimulationLanguageOverride_jaToEn() async throws {
    // Scenario: ja authoring, en simulation override. prompt:nil forces fallback.
    // The captured prompt must contain the English fallback, not the Japanese one.
    let mock = MockLLMService(responses: [
      #"{"vote": "Bob"}"#,
      #"{"vote": "Alice"}"#
    ])
    try await mock.loadModel()

    let scenario = ScenarioFixture.make(
      language: "ja",
      simulationLanguage: "en",
      personas: [
        Persona(name: "Alice", description: "A test persona"),
        Persona(name: "Bob", description: "A test persona")
      ],
      phases: [Phase(type: .vote, prompt: nil, outputSchema: ["vote": "string"])]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    let prompt = mock.capturedPrompts[0].user
    #expect(prompt.contains("Vote for the person"))
    #expect(!prompt.contains("最も怪しい"))
  }

  @Test func voteHonorsSimulationLanguageOverride_enToJa() async throws {
    // Reverse: en authoring, ja simulation override. Captured prompt must contain Japanese.
    let mock = MockLLMService(responses: [
      #"{"vote": "Bob"}"#,
      #"{"vote": "Alice"}"#
    ])
    try await mock.loadModel()

    let scenario = ScenarioFixture.make(
      language: "en",
      simulationLanguage: "ja",
      personas: [
        Persona(name: "Alice", description: "A test persona"),
        Persona(name: "Bob", description: "A test persona")
      ],
      phases: [Phase(type: .vote, prompt: nil, outputSchema: ["vote": "string"])]
    )
    var state = SimulationState.initial(for: scenario)
    state.currentRound = 1
    let collector = EventCollector()

    let context = makePhaseContext(scenario: scenario, llm: mock, collector: collector)
    try await handler.execute(context: context, state: &state)

    let prompt = mock.capturedPrompts[0].user
    #expect(prompt.contains("最も怪しい"))
    #expect(!prompt.contains("Vote for the person"))
  }
}

File: shared/engine/src/commonTest/kotlin/com/pastura/engine/VoteHandlerTests.kt
package com.pastura.engine

import com.pastura.models.Persona
import com.pastura.models.Phase
import com.pastura.models.PhaseType
import com.pastura.models.Scenario
import com.pastura.models.SimulationEvent
import com.pastura.models.SimulationState
import com.pastura.models.TurnOutput
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Kotlin sibling of Swift's `VoteHandlerTests` (+ its `…+TurnDegradation` split),
 * collapsed into one class per the commonTest convention.
 *
 * Vote is a hybrid: it WRITES `lastOutputs` (and clears a stale entry on skip, like
 * [SpeakAllHandler]) but never appends to `conversationLog` (a vote is not an
 * utterance). Off-candidate votes are dropped from the tally yet kept in the raw
 * `votes` map (#524). The ADR-021 turn-gate transient-skip is exercised here; the
 * D3 systemic-error and D4 circuit-breaker cases ride the same shared gate already
 * covered by [SpeakAllHandlerTests] + [TurnFailureGateTests].
 *
 * Ported for the ADR-023 KMP Engine migration (#501, #1249).
 */
class VoteHandlerTests {

    private val handler = VoteHandler()

    private fun scenario(
        agents: List<String> = listOf("Alice", "Bob", "Charlie"),
        language: String = "en",
        simulationLanguage: String? = null,
        prompt: String? = "Vote!",
        excludeSelf: Boolean? = true,
        outputSchema: Map<String, String> = mapOf("vote" to "string"),
    ) = Scenario(
        id = "t",
        name = "T",
        description = "d",
        language = language,
        simulationLanguage = simulationLanguage,
        agentCount = agents.size,
        rounds = 2,
        logWindow = null,
        context = "A test.",
        personas = agents.map { Persona(name = it, description = "$it's persona.") },
        phases = listOf(
            Phase(type = PhaseType.VOTE, prompt = prompt, outputSchema = outputSchema, excludeSelf = excludeSelf),
        ),
    )

    private fun context(
        scenario: Scenario,
        backend: LLMBackend,
        events: MutableList<SimulationEvent> = mutableListOf(),
    ) = PhaseContext(
        scenario = scenario,
        phase = scenario.phases[0],
        backend = backend,
        suspensionRelay = SuspensionRelay(),
        emitter = { events += it },
        pauseCheck = { },
        phasePath = listOf(0),
        turnGate = TurnFailureGate(),
    )

    private fun votes(target: String) =
        ScriptedLLMBackend.Script.completing("""{"vote": "$target"}""")

    private fun voteResultsEvent(events: List<SimulationEvent>) =
        events.filterIsInstance<SimulationEvent.VoteResults>().single()

    private fun initial(s: Scenario) = SimulationState.initial(s).copy(currentRound = 1)

    // MARK: - Core tally

    @Test
    fun collectsVotesFromAllAgents() = runTest {
        val s = scenario()
        val backend = ScriptedLLMBackend(listOf(votes("Bob"), votes("Alice"), votes("Alice")))
        val next = handler.execute(context(s, backend), initial(s))

        assertEquals(2, next.voteResults["Alice"])
        assertEquals(1, next.voteResults["Bob"])
        assertEquals(3, backend.callCount)
    }

    @Test
    fun emitsVoteResultsEvent() = runTest {
        val s = scenario(agents = listOf("Alice", "Bob"))
        val events = mutableListOf<SimulationEvent>()
        val backend = ScriptedLLMBackend(listOf(votes("Bob"), votes("Alice")))
        handler.execute(context(s, backend, events), initial(s))

        val event = voteResultsEvent(events)
        assertEquals("Bob", event.votes["Alice"])
        assertEquals("Alice", event.votes["Bob"])
    }

    @Test
    fun skipsEliminatedAgents() = runTest {
        val s = scenario()
        // Bob eliminated → Alice and Charlie vote; both target the other active pick.
        val backend = ScriptedLLMBackend(listOf(votes("Charlie"), votes("Alice")))
        val state = initial(s).copy(eliminated = mapOf("Bob" to true))
        handler.execute(context(s, backend), state)

        assertEquals(2, backend.callCount)
    }

    @Test
    fun populatesVoteResultsStateVariable() = runTest {
        // Key must be "vote_results" (plural) to match the {vote_results} placeholder
        // documented in PhaseEditorSheet and used by the word_wolf preset.
        val s = scenario()
        val backend = ScriptedLLMBackend(listOf(votes("Bob"), votes("Alice"), votes("Alice")))
        val next = handler.execute(context(s, backend), initial(s))

        assertEquals("""{"Alice": 2, "Bob": 1}""", next.variables["vote_results"])
        assertNull(next.variables["vote_result"])
    }

    // MARK: - Off-candidate tally drop (#524)

    @Test
    fun dropsInvalidVotesFromTally() = runTest {
        // Three voters (exclude_self default true):
        //   Alice → "Alice"   self-vote, invalid (Alice ∉ her candidates)
        //   Bob   → "Ghost"   hallucinated name, invalid
        //   Charlie → "Bob"   valid
        // Reverting the `candidates.contains(votedFor)` guard would put Alice (self)
        // and Ghost back into the tally — this test goes red in that case.
        val s = scenario()
        val events = mutableListOf<SimulationEvent>()
        val backend = ScriptedLLMBackend(listOf(votes("Alice"), votes("Ghost"), votes("Bob")))
        val next = handler.execute(context(s, backend, events), initial(s))

        assertEquals(mapOf("Bob" to 1), next.voteResults)
        assertNull(next.voteResults["Alice"]) // self-vote dropped
        assertNull(next.voteResults["Ghost"]) // hallucinated dropped

        // Divergence is intentional: the raw votes stay visible in the VoteResults
        // event even though they are absent from the tally.
        val event = voteResultsEvent(events)
        assertEquals("Alice", event.votes["Alice"]) // self-vote preserved in votes map
        assertEquals("Ghost", event.votes["Bob"]) // hallucinated preserved in votes map
        assertEquals("Bob", event.votes["Charlie"])
    }

    @Test
    fun dropsVotesForEliminatedAgentsFromTally() = runTest {
        // Disjoint negative control for the `state.eliminated[name] == true` branch of
        // voteCandidates (the #524 self-vote test above covers only the exclude_self
        // branch). Charlie eliminated; Alice votes the eliminated Charlie (invalid —
        // Charlie ∉ Alice's candidates [Bob]), Bob votes Alice (valid). Reverting the
        // eliminated filter would make Charlie a candidate → Alice's vote tallied → red.
        val s = scenario()
        val events = mutableListOf<SimulationEvent>()
        val backend = ScriptedLLMBackend(listOf(votes("Charlie"), votes("Alice")))
        val state = initial(s).copy(eliminated = mapOf("Charlie" to true))
        val next = handler.execute(context(s, backend, events), state)

        assertEquals(mapOf("Alice" to 1), next.voteResults)
        assertNull(next.voteResults["Charlie"]) // vote for an eliminated agent dropped

        // Raw value still observable in the votes map.
        assertEquals("Charlie", voteResultsEvent(events).votes["Alice"])
    }

    // MARK: - Hybrid write semantics (writes lastOutputs, never conversationLog)

    @Test
    fun writesBallotToLastOutputsAndLeavesConversationLogUntouched() = runTest {
        val s = scenario(agents = listOf("Alice", "Bob"))
        val backend = ScriptedLLMBackend(listOf(votes("Bob"), votes("Alice")))
        val before = initial(s)
        val next = handler.execute(context(s, backend), before)

        // lastOutputs written (relationship_update reads lastOutputs[voter].vote).
        assertEquals("Bob", next.lastOutputs["Alice"]?.fields?.get("vote"))
        assertEquals("Alice", next.lastOutputs["Bob"]?.fields?.get("vote"))
        // A vote is not an utterance — never appended to the conversation log.
        assertTrue(next.conversationLog.isEmpty())
        // Input untouched (immutable-state contract).
        assertTrue(before.lastOutputs.isEmpty())
    }

    // MARK: - simulationLanguage override (ADR-010 Step E)

    @Test
    fun voteHonorsSimulationLanguageOverrideJaToEn() = runTest {
        // ja authoring, en simulation override, prompt:null forces the fallback. The
        // captured prompt must contain the English fallback, not the Japanese one.
        val s = scenario(agents = listOf("Alice", "Bob"), language = "ja", simulationLanguage = "en", prompt = null)
        val backend = ScriptedLLMBackend(listOf(votes("Bob"), votes("Alice")))
        handler.execute(context(s, backend), initial(s))

        val prompt = backend.requests[0].user
        assertTrue(prompt.contains("Vote for the person"))
        assertTrue(!prompt.contains("最も怪しい"))
    }

    @Test
    fun voteHonorsSimulationLanguageOverrideEnToJa() = runTest {
        // Reverse: en authoring, ja simulation override.
        val s = scenario(agents = listOf("Alice", "Bob"), language = "en", simulationLanguage = "ja", prompt = null)
        val backend = ScriptedLLMBackend(listOf(votes("Bob"), votes("Alice")))
        handler.execute(context(s, backend), initial(s))

        val prompt = backend.requests[0].user
        assertTrue(prompt.contains("最も怪しい"))
        assertTrue(!prompt.contains("Vote for the person"))
    }

    // MARK: - Prompt wiring (candidates injection)

    @Test
    fun candidatesVariableIsExpandedForEachVoter() = runTest {
        // The {candidates} variable is the voter's candidate list (exclude_self
        // default true), so Alice's prompt lists Bob, Charlie — never herself.
        val s = scenario(prompt = "Candidates: {candidates}")
        val backend = ScriptedLLMBackend(listOf(votes("Bob"), votes("Alice"), votes("Alice")))
        handler.execute(context(s, backend), initial(s))

        assertEquals("Candidates: Bob, Charlie", backend.requests[0].user)
    }

    // MARK: - captureMood round-trip (single-copy fold, #913)

    @Test
    fun capturedMoodSurfacesInTheNextRoundsPrompt() = runTest {
        // Pins the single success-path `state.copy` fold: captureMood must land in
        // the RETURNED state (folded alongside lastOutputs), else round N's mood never
        // reaches round N+1. Perturbing the captureMood fold turns this red.
        val moodPhase = Phase(
            type = PhaseType.VOTE,
            prompt = "Vote!",
            outputSchema = mapOf("vote" to "string", "mood" to "string"),
        )
        val s = scenario(agents = listOf("Alice", "Bob")).copy(phases = listOf(moodPhase))
        val backend = ScriptedLLMBackend(
            listOf(
                ScriptedLLMBackend.Script.completing("""{"vote": "Bob", "mood": "わくわく"}"""),
                ScriptedLLMBackend.Script.completing("""{"vote": "Alice", "mood": "冷静"}"""),
                ScriptedLLMBackend.Script.completing("""{"vote": "Bob", "mood": "不安"}"""),
                ScriptedLLMBackend.Script.completing("""{"vote": "Alice", "mood": "焦り"}"""),
            ),
        )
        val ctx = context(s, backend)

        val afterRound1 = handler.execute(ctx, initial(s))
        assertEquals("わくわく", afterRound1.variables["mood_Alice"])

        handler.execute(ctx, afterRound1)
        // Alice's round-2 prompt (request index 2) must carry her captured mood.
        assertTrue(backend.requests[2].system.contains("わくわく"))
        assertTrue(backend.requests[2].system.contains("Your Current Mood"))
    }

    // MARK: - Empty vote is a skip, NOT a tally drop (parser-guard distinction)

    @Test
    fun emptyVoteIsAbsorbedAsSkipNotTallyDrop() = runTest {
        // The parser rejects a present-but-empty expected key ({"vote":""} —
        // `hasAllExpectedKeys` requires non-empty content), so three empties exhaust
        // the retry budget → RetriesExhausted → the turn gate absorbs it as a skip
        // (ADR-021 D2). It is therefore an ABSTENTION (no ballot, stale lastOutputs
        // cleared), NOT an off-candidate tally drop — the two paths must not be
        // conflated. Asserted through the skip mechanism, not a tally value.
        val s = scenario(agents = listOf("Alice"))
        val backend = ScriptedLLMBackend(listOf(votes(""), votes(""), votes("")))
        val events = mutableListOf<SimulationEvent>()
        val state = initial(s).copy(
            lastOutputs = mapOf("Alice" to TurnOutput(fields = mapOf("vote" to "stale"))),
        )
        val next = handler.execute(context(s, backend, events), state)

        // Abstention: no ballot in the tally or the votes map, stale output cleared.
        assertTrue(next.voteResults.isEmpty())
        assertNull(voteResultsEvent(events).votes["Alice"])
        assertNull(next.lastOutputs["Alice"])

        // Mechanism: absorbed as a turn skip (no AgentOutput), not a guard-dropped tally.
        assertTrue(events.filterIsInstance<SimulationEvent.AgentOutput>().isEmpty())
        val skipped = events.filterIsInstance<SimulationEvent.TurnSkipped>()
        assertEquals(1, skipped.size)
        assertEquals("Alice", skipped.single().agent)
        assertEquals(PhaseType.VOTE, skipped.single().phaseType)
    }

    // MARK: - Turn degradation (ADR-021 D1/D2)

    @Test
    fun transientFailureAbstainsAndOthersStillVote() = runTest {
        // Alice's call fails transiently (turn-degradable); Bob and Charlie still
        // vote. Alice's ballot is absent from both the votes map and the tally
        // (abstention), and her stale lastOutputs entry is cleared (ADR-021 D2).
        val s = scenario()
        val backend = ScriptedLLMBackend(
            listOf(
                ScriptedLLMBackend.Script(terminal = TerminalStatus.Failed(errorCode = "transient blip")),
                votes("Alice"),
                votes("Alice"),
            ),
        )
        val events = mutableListOf<SimulationEvent>()
        val before = initial(s).copy(
            lastOutputs = mapOf("Alice" to TurnOutput(fields = mapOf("vote" to "stale"))),
        )
        val next = handler.execute(context(s, backend, events), before)

        // Only Bob and Charlie's ballots land; both voted "Alice".
        assertEquals(2, next.voteResults["Alice"])
        assertEquals(1, next.voteResults.size)

        // No AgentOutput for the skipped voter; exactly one TurnSkipped.
        val outputs = events.filterIsInstance<SimulationEvent.AgentOutput>().map { it.agent }
        assertEquals(listOf("Bob", "Charlie"), outputs)

        val skipped = events.filterIsInstance<SimulationEvent.TurnSkipped>()
        assertEquals(1, skipped.size)
        assertEquals("Alice", skipped.single().agent)
        assertEquals(PhaseType.VOTE, skipped.single().phaseType)

        // Abstention: no ballot for Alice in the emitted votes map, stale output cleared.
        val event = voteResultsEvent(events)
        assertNull(event.votes["Alice"])
        assertEquals("Alice", event.votes["Bob"])
        assertNull(next.lastOutputs["Alice"])
        assertEquals("Alice", next.lastOutputs["Bob"]?.fields?.get("vote"))
    }
}
